lib/absinthe/federation/schema/entity_union.ex

```diff
@@ -64,19 +67,41 @@ end
 defimpl Absinthe.Federation.Schema.EntityUnion.Resolver, for: Any do
   alias Absinthe.Adapter.LanguageConventions
 
-  def resolve_type(%struct_name{}, resolution) do
-    struct_name
-    |> Module.split()
-    |> List.last()
-    |> to_internal_name(resolution.adapter)
+  def resolve_type(%struct_name{} = data, resolution) do
+    typename =
+      struct_name
+      |> Module.split()
+      |> List.last()
+
+    inner_resolve_type(data, typename, resolution)
+  end
+
+  def resolve_type(%{__typename: typename} = data, resolution) do
+    inner_resolve_type(data, typename, resolution)
   end
 
-  def resolve_type(%{__typename: typename}, resolution) do
-    to_internal_name(typename, resolution.adapter)
+  def resolve_type(%{"__typename" => typename} = data, resolution) do
+    inner_resolve_type(data, typename, resolution)
   end
 
-  def resolve_type(%{"__typename" => typename}, resolution) do
-    to_internal_name(typename, resolution.adapter)
+  defp inner_resolve_type(data, typename, resolution) do
+    type = Absinthe.Schema.lookup_type(resolution.schema, typename)
+
+    if is_nil(type) do
+      to_internal_name(typename, resolution.adapter)
+    else
+      if(Map.has_key?(type, :resolve_type)) do
+        resolver = Absinthe.Type.function(type, :resolve_type)
+
+        if is_nil(resolver) do
+          to_internal_name(typename, resolution.adapter)
+        else
+          resolver.(data, resolution)
+        end
+      else
+        to_internal_name(typename, resolution.adapter)
+      end
+    end
```

Comment:
nice logic here - may I suggest the following pattern to simplify it a bit?

```suggestion
    case type do
      %{resolve_type: resolve_type} when not is_nil(resolve_type) ->
        resolver = Absinthe.Type.function(type, :resolve_type)
        resolver.(data, resolution)

      _type ->
        to_internal_name(typename, resolution.adapter)
    end
  end
```

I believe `:resolve_type` must carry a function ref if it's not `nil`, so we shouldn't need the extra check whether `Absinthe.Type.function/2` is returning an actual function.